src/builder/specification.md

```diff
@@ -0,0 +1,314 @@
+    eth1_data: Eth1Data
+    graffiti: Bytes32
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: SyncAggregate
+    execution_payload_header: ExecutionPayloadHeader
+```
+
+## Errors
+
+The list of error codes introduced by this specification can be found below.
+
+| Code | Message | Meaning |
+| - | - | - |
+| -32000 | Server error | Generic client error while processing request. |
+| -32001 | Unknown hash | No block with the provided hash is known. |
+| -32002 | Unknown validator | Unknown validator. |
+| -32003 | Unknown fee recipient | No known mapping between validator and fee recipient. |
+| -32004 | Unknown block | Block does not match the provided header. |
+| -32005 | Invalid signature | Provided signature is invalid. |
+| -32006 | Invalid timestamp | Provided timestamp was invalid. |
+| -32600 | Invalid request | The JSON sent is not a valid Request object. |
+| -32601 | Method not found | The method does not exist / is not available. |
+| -32602 | Invalid params | Invalid method parameter(s). |
+| -32603 | Internal error | Internal JSON-RPC error. |
+| -32700 | Parse error | Invalid JSON was received by the server. |
+
+Each error returns a `null` `data` value, except `-32000` which returns the `data` object with a `err` member that explains the error encountered.
+
+For example:
+
+```
+$ curl https://localhost:8550 \
+    -X POST \
+    -H "Content-Type: application/json" \
+    -d '{"jsonrpc":"2.0","method":"engine_getPayloadV1","params": ["0x1"],"id":1}'
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "error": {
+    "code": -32000,
+    "message": "Server error",
+    "data": {
+        "err": "Database corrupted"
+    }
+  }
+}
+```
+
+## Routines
+
+### Signing
+
+All signature operations should follow the [standard BLS operations][bls] interface defined in `consensus-specs`.
+
+There are two types of data to sign over in the Builder API:
+* In-protocol messages, e.g. [`BlindedBeaconBlock`](#blindedbeaconblock), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and use the domain specified for beacon block proposals.
+* Builder API messages, e.g. [`builder_registerValidatorV1`](#builder_registerValidatorV1) and the response to [`builder_getHeader`](#response-2), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and the domain `DomainType('0xXXXXXXXX')` (TODO: get a proper domain).
+
+As `compute_signing_root` takes `SSZObject` as input, client software should convert in-protocol messages to their SSZ representation to compute the signing root and Builder API messages to the SSZ representations defined [above](#sszobjects).
+
+## Methods
+
+### `builder_status`
```

Comment:
clarification:

does this endpoint serve as a liveness check for the `mev_boost` server? or the upstream relays?

Reply:
Agree, this should be clarified, at least in the spec with an explanatory sentence! From the context I had on this, it was for only querying status of the `mev_boost` server, which is also the current implementation (just checks whether it's alive).

It could be extended to the relays and return OK if at least one relay returns ok, if that would be more useful.

Reply:
Thinking about it, I think maybe this call should return an object that can later be extended with relay liveness in a backward compatible manner.

Reply:
yeah agreed, my understanding was just to ping the local `mev-boost` server to ensure it is up

can extend down the line to check for relay liveness although that seems like the kind of thing that should almost just be internal to the `mev-boost` engine so validator clients don't have to bother with it

Reply:
I feel like it is an implementation detail of `mev-boost` on how it should respond. I don't think this endpoint should be extended to understand the concept of multiple relays. IMO the simplicity of this comes from assuming `mev-boost` multiplexes the responses. `mev-boost` should decided what number of relays it needs an `"OK"` from to be confident it will be able to produce a block.